modules/xcm-bridge-hub/src/lib.rs

```diff
@@ -0,0 +1,1727 @@
+			log::trace!(
+				target: LOG_TARGET,
+				"Bridge {:?} between {:?} and {:?} has been opened",
+				locations.lane_id,
+				locations.bridge_origin_universal_location,
+				locations.bridge_destination_universal_location,
+			);
+
+			// deposit `BridgeOpened` event
+			Self::deposit_event(Event::<T, I>::BridgeOpened {
+				lane_id: locations.lane_id,
+				local_endpoint: Box::new(locations.bridge_origin_universal_location),
+				remote_endpoint: Box::new(locations.bridge_destination_universal_location),
+			});
+
+			Ok(())
+		}
+
+		/// Request previously opened bridge closure.
+		///
+		/// Can only be called by the "owner" of this side of the bridge. Both inbound and outbound
+		/// lanes at this side of the bridge are immediately switched to `Closing` state. Lanes may
+		/// still send/receive new messages, however this state is reported to the other side of
+		/// the bridge, given that relays are active. When this happens, lanes at the other side
+		/// also switching their state to `Closing`.
+		///
+		/// After `T::BridgeCloseDelay` blocks, the caller may issue the `close_bridge` call to
+		/// actually close the bridge and claim reserved funds.
+		///
+		/// The states after this call: bridge is `Closing` in `BridgeCloseDelay` blocks, outbound
+		/// lane is `Closing`, inbound lane is `Closing`.
+		#[pallet::call_index(1)]
+		#[pallet::weight(Weight::zero())] // TODO: https://github.com/paritytech/parity-bridges-common/issues/1760 - weights
+		pub fn request_bridge_closure(
+			origin: OriginFor<T>,
+			bridge_destination_relative_location: MultiLocation, // TODO: versioned?
+		) -> DispatchResult {
+			// compute required bridge locations
+			let locations = Self::bridge_locations(origin, bridge_destination_relative_location)?;
+
+			// update bridge metadata
+			let may_close_at =
+				Self::start_closing_the_bridge(locations.lane_id, false, LaneState::Closing)?;
+
+			// write something to log
+			log::trace!(
+				target: LOG_TARGET,
+				"Bridge {:?} between {:?} and {:?} is closing. May be closed at: {:?}",
+				locations.lane_id,
+				locations.bridge_origin_universal_location,
+				locations.bridge_destination_universal_location,
+				may_close_at,
+			);
+
+			// deposit the `BridgeClosureRequested` event
+			Self::deposit_event(Event::<T, I>::BridgeClosureRequested {
+				lane_id: locations.lane_id,
+				may_close_at,
+			});
+
+			Ok(())
+		}
+
+		/// Try to close the bridge.
+		///
+		/// Can only be called by the "owner" of this side of the bridge. Can only be called if
+		/// bridge has switched to `Closing` state at least `T::BridgeCloseDelay` blocks ago.
+		///
+		/// Closed bridge is a bridge without any traces in the runtime storage. So this method
+		/// first tries to prune all queued messages at the outbound lane. When there are no
+		/// outbound messages left, outbound and inbound lanes are pruned. After that, funds
+		/// are returned back to the owner of this side of the bridge.
```

Comment:
Yes, that would great. Let me think about it a bit more. I'll explain why right now they're simply pruned:
1) there most probably will be a misbehaviour for the case when messages are not relayed. I.e. if message `N` has been queued a week/month ago, then bridge has no relayers = misbehavior. So if we'll be waiting for messages to be realyed on such a bridge, we may wait forever;
2) there will be some relayers coordination protocol in the future. And if message will be closing, then probably the bridge owner will stop to fund relayers at some point (during `Closing` period). So messages again won't be relayed.

But I think we may improve closing procedure a bit - maybe we should allow other actors to call `close_bridge` (or some sibling function) with proof of relaying some messages => draining reserved funds from the owner sovereign account.

Also I think there's another problem with that - I've been assuming that after proper closing the bridge may be reopened again (probably with some higher reserve - not implemented now). But what if at one side of the bridge the bridge has already been closed (vanished from the storage) and at the other side it still in the `Closed` phase, that needs pruning. That's definitely a miscoordination between bridge endpoints (they should talk offchain and make offchain decisions about that), but it may lead to some issues - e.g. when messages from the "old bridge" are delivered to the "new bridge". I think we may imrpove it as well

That said - let me think about it a bit. It isn't trivial. Thanks for suggestion

Reply:
I was battling myself for several hours and here's my updated vision:
1) current closing procedure was meant to introduce some kind of "graceful shutdown" for the bridge. So one side is closing the bridge, then another side gets notification and also starts closing the bridge. They have enough time to send and deliver required messages, but if they fail to do that, we just prune remaining messages to make sure they don't spoil the bridge hub storage;
2) let's only consider "regular" closure here - i.e. won't touch misbehavior. So what are possible reasons for the bridge closure? IMO: some kind of a break up between two chains, detected cheating by the other side, merging/migration to the other chain and so on. All those reasons are outside of the bridge hub scope and chains will need to explicitly talk to each other and their users (who have e.g. sent funds over the bridge). And our "graceful closure" period is just a helper here - it won't help to solve all the issues, because they are not external to the bridge hub
3) both connected chains may just update their configuration and discard messages during closure period, so again - we're unneeded here.

So it seems that "regular" closure process shall be opaque to the bridge hub itself - that's a task for connected chains to organize that properly, probably they will solve it without using the bridge. So: this graceful shutdown is not needed at all. Bridges must be closed immediately, no matter what happens at the other side of the bridge.

---

There's another kind of closure - the outcome of misbehavior. In this case we also currently has this graceful closure, which works only for the bridged chain, though. I thought that the bridged chain may send messages, required e.g. to unreserve funds of its accounts. OTOH the other chain may use this as an advantage - i.e. it may spend some money to flood bridge queue at the bridged side, thus closing the bridged outbound queue. Then it will be able to send its own messages, while the other side won't.

I have no idea how to handle misbehavior more accurate. Like if we'll be ignoring that (for example by ripping some portion of reserved funds once per day), in the end it may lead to enormous storage growth or something like that. Most of misbehaviors will need a time to solve and we can't leave bridge running during that time. So misbehavior is something that we can't allow. So I'm tending to:
1) immediately close both inbound and outbound lanes;
2) allow reporter to prune queued messages and get all the reserved funds once he'll do that;
3) do not allow to reopen bridge - i.e. leave some trace in the runtime storage forever (relay chain governance may remove this entry later, though).

This way we will force bridge owners to watch their queues state accurately and take care of that. WDYT?

Reply:
Seems like a tough problem. I would like to spend some more time thinking about it. I also need to do a more in-depth iteration on the PR, and maybe look on #2233. I'll get back as soon as I get a better understanding of everything.

Reply:
Yep, thank you! I think we may only adjust something in the second part, related to misbehavior. The first part is what I'm going to change soon

Reply:
I've removed the `Closing` state, so now when someone (owner or misbehavior reporter) closes the bridge, it happens immediately. The queued messages must be pruned as before - for regular closure I assume that there won't be any messages (it must be organized by the connected parties on their level - e.g. by closing our router some time before bridge is closed).

For misbehavior closure the process is the same, however I'm also waiting for your opinion + still thinking myself about benefits of changing it to something that may be better for bridge users.

Reply:
Another idea that I've had over weekend: instead of closing the lanes, we can temporary close HRMP/DMP channel with the sending chain OR simply stop processing messages from this chain. This way we: (1) will fine the bridge owner for misbehavior (2) stop pushing new messages to the bridge queue until e.g. this queue is empty (3) do not lose any messages (4) messages will keep piling up at the sending chain => that's now a sending chain problem and it doesn't spoil the bridge hub queues.

This would require some changes in the XCMP/DMP pallets, but if we are sure that that's the best option - we can go on with that

Reply:
I like this idea of closing/suspending HRMP/DMP channel

Reply:
I also like the idea of closing HRMP/DMP channel,
there is some handling for hrmp in xcm (unimplemented):
```
                        HrmpNewChannelOpenRequest { .. } => Err(XcmError::Unimplemented),
			HrmpChannelAccepted { .. } => Err(XcmError::Unimplemented),
			HrmpChannelClosing { .. } => Err(XcmError::Unimplemented),
```

if we could get it working like:
1. close hrmp channel on BridgeHub
2. which could bubble to relay chain and "close" stuff there
3. which could bubble as `HrmpChannelClosing` to sibling parachain and "close" stuff there
the result would be that sibling chain's XcmRouter will fail after this (and needs to go through governance again to open hrmp...)

well yes, there is some window between trigger close_hrmp and real closing on sibling, but probably we just process messages on the way or whatever policy



we could discuss this with XCM team, for example system-parachain is kind of "relay-chain" extension, so to use this approach could be ok...
```
        ...
        ...
        /// Safety: The message should originate directly from the relay-chain.
	///
	/// Kind: *System Notification*
	///
	/// Errors:
	HrmpChannelClosing {
```

Reply:
This is now fixed. The XCM channel operations are now hidden behind `LocalChannelManager` trait - I'll implement it later in the #2255, because it needs separate investigation + there's an ongoing refactoring of XCM queues, which may affect the implementation. Other than that, it'll now works as follows:
1) `misbehavior_report` leads to immediate inbound XCM channel closure, meaning that no inbound messages can be enqueued into the bridge queue. Outbound lane is now `Closed`
2) I assume that all misbehavior types are about proper processing of bridge queues, so the owner/relayer must be able to deliver pending messages to the bridged side even in this state (there's a pending TODO in the messages pallet - I'll extract it into a separate PR);
3) once it happens, he may use some **other** account at bridge hub to resume the bridge. It resumes all misbehaving bridges of the owner at the same time, so shall be called when all bridges are "fixed". This account replenish the `Penalty` on the bridge owner account and this amount gets reserved there.